### DISCUSSION TITLE
Panic: buffer is empty - Merge block is not working as expected

### DISCUSSION THREAD

**Discussion by BrianJonesNMI** [postmanlabs/postman-flows#146] [Q&A]
I have an OAuth flow. I get the token from a previous request, pass that response to an Assign Variables then pass the token to a merge block. the same token is passed to the next request and I am getting a list that I pass to a ForEach and the ForEach passes an ID to an assign variable then to the previously mentioned Merge.

If either the assign variable or the for each block passes the data to the target connector I get "Panic: buffer is empty" message in the lower right-hand corner and nothing in the console or terminal.

If I attach both blocks to the source connector the flow just stops. Nothing logged to terminal or console.

When I changed the Assign Variable from target to source I get an $ Unknown which I can not seem to change and the output is empty.

One of the variables is a hardcoded value at the moment I removed the hardcoded value and the results were the same

I am unsure how to proceed here. any assistance would be appreciated.

![BothSourced](https://user-images.githubusercontent.com/44819977/144328376-8575b432-8965-4b40-a2d9-4e1fbd7990a3.png)

**Comment by saswatds** [comment]
@BrianJonesNMI If understand this correctly, you are trying to persist the access token across all the different requests which you want to send next. Correct?

The `Panic: Buffer is empty` seems to be a bug in the runtime. I will get a fix out quickly.  But having said that, this flow would not work how you intended it to work. Each block generates data in discrete packets and they keep moving forward. 

The first `Assign variable` block will just generate 1 packet of data containing Token and the number `6953`. And it only gets merged with the first element that was obtained for `ForEach`. 

Until now Flows does not have a global state and this is the root cause of the problem. We are working on a way using which you will be able to create some data that will be transparently passed across to all blocks similar to how a global state works. I will keep you posted when that is out.

**Comment by BrianJonesNMI** [comment]
Thanks for responding.

you state: 
The first Assign variable block will just generate 1 packet of data containing Token and the number 6953. And it only gets merged with the first element that was obtained for ForEach.

This is not happening I either get a Panic error or nothing, how do I even get the first request from the for-each to make a request?

So what you are saying is in the Flows currently there is no way to persist a token across multiple requests. So in order to do loop around an OAuth service I have to get a new token every request?

Has anyone solved this issue?

**Reply by saswatds** [reply]
@BrianJonesNMI I am debugging this issue personally, I will ensure a hotfix is out as soon as I fix it. Anything other than a persisted state would cause a very messy flow. I was able to implement something like a global state within the runtime. I will push to production after performing some sanity testing.

**Reply by BrianJonesNMI** [reply]
@saswatds Thanks for doing the work yourself. Do you have any updates?

Thanks

**Reply by saswatds** [reply]
Hey @BrianJonesNMI, you should now able to see a new block called `Create Durables`. Here is also a small show and tell video explaining the same https://github.com/postmanlabs/postman-flows/discussions/171. You can use this to set your access token in the durable scope and use it without needing to merge. This will also remove the Panic because merge will not try to combine non-existent data

**Reply by BrianJonesNMI** [reply]
Thank you what version of the desktop will this feature be included in. I can not use the web version as most of my work is behind a firewall and the last time i tried the agent it was not working, too many beta tools being stacked together, for my preference.

**Reply by BrianJonesNMI** [reply]
I found it in the current version. Thanks
